content/11-serverless/03-serverless-challenges.mdx

```diff
@@ -0,0 +1,105 @@
+
+These solutions start to blur the line a bit as to what constitutes a serverless environment.  When developers are forced to pay for standby resources when their code is not actively executing, it raises some questions about some of the fundamental claims of the serverless paradigm.
+
+A recent alternative to preallocating resources has been to sidestep the problem by switching to a lighter runtime environment.  Runtimes like V8 have a much different execution strategy than traditional serverless and are able to avoid cold start problems by using different isolation technologies and a more pared down environment.  They avoid cold start issues at the expense of compatibility with functions that have dependencies on a more robust environment.
+
+## Application design constraints
+
+Another challenge that is fundamental to the serverless model is the application design it imposes.  Serverless platforms are only useful for applications that can work within their constraints.  Some of these are inherent in cloud computing in general, while other requirements are dictated by the serverless model specifically.
+
+### Designing a cloud-friendly architecture
+
+The first requirement that applications must meet to use serverless platforms is to be designed in a cloud-friendly way.  In the context of this discussion, at the very least this means that the application must be at least partly deployable to a cloud service that the other components are able to communicate with.  And while it's possible to implement monolithic functions in your design, the serverless model best accommodates microservice architectures.
+
+The upshot of this is that your application must be designed in part as a series of functions executed by your serverless provider.  You must be comfortable with the processing taking place on infrastructure that you do not control.  Furthermore, you must be able to decompose your application's functionality into discrete functions that can be executed remotely.
+
+### Dealing with stateless execution
+
+Serverless functions are, by design, stateless.  That means that, while some information may possibly be cached if the function is executed with the same resources, you can't rely on any state being shared between invocations of your functions.
+
+You must design your functions to have all of the information they need to execute internally.  Any external state must be fetched at the beginning of invocation and exported before finishing.  Since functions may be executed in parallel, this also limits what type of state may reasonable acted upon.  In general, the less state that your functions have to manage, the faster and cheaper they will be to execute and the less complexity you will have to manage.
```

Comment:
"the less states". It might be correct the way you wrote it, and I am just unfamiliar with the way this is said in the plural.

Reply:
I had to think about this one for awhile upon your comment here.  In this case, I think "the less state" is actually in line with how I've seen other people talk about it.  I can't tell you if it's grammatical, but I do think that it matches the way I've seen people use it.  I think that might be because in these cases, developers talk about "state" as a kind of ambiguously sized glob.  I'm not sure though.

Reply:
That's what I was assuming, but I wasn't sure.